Namespace-specified imports

adamgundry opened the pull request:
The [proposal](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0581-namespace-specified-imports.rst) has been accepted; the following discussion is mostly of historic interest.

--------------------------------------------

This proposal allows imports to be limited to either the type or data namespace. Moreover it clarifies the specification of
`ExplicitNamespaces` and related extensions. This is heavily derived from #270. Thanks to @Hithroc and @cdornan for collaborating on this! This also includes changes originally prepared by @int-index amending proposal #65.

[Rendered](https://github.com/adamgundry/ghc-proposals/blob/namespace-specified-imports/proposals/0000-namespace-specified-imports.rst).
[Work-in-progress user's guide section with these changes](https://github.com/cdornan/ExplicitNamespaces-doc/pull/1).

int-index committed:
Amend value->data in proposal 65

adamgundry committed:
Namespace-specified imports proposal

This proposal allows imports to be limited to either the type or
data namespace. Moreover it clarifies the specification of
ExplicitNamespaces and related extensions.

Co-authored-by: Artyom Kuznetsov <hi@wzrd.ht>
Co-authored-by: Chris Dornan <chris@chrisdornan.com>

nomeata commented on proposals/0000-namespace-specified-imports.rst:
```diff
@@ -0,0 +1,554 @@
+  imports, we could introduce a new extension ``NamespacedImports``.  This would
+  make it clearer whether code depends on the new feature, or whether the older
+  version of ``ExplicitNamespaces`` was enough.  However the general consensus
+  seems to favour reducing the number of extensions over avoiding change to
```
That's debatable :-). We'll know more once we have concluded the ongoing policy discussion. But since we didn't add ExplicitNamespaces to GHC2023, likely fine here

int-index commented on proposals/0000-namespace-specified-imports.rst:
```diff
@@ -0,0 +1,554 @@
+     , T(pattern D) -- Rejected: pattern keyword cannot be used in sub-list
+     , data T       -- Rejected: T not in scope in data namespace
+     , type E       -- Rejected: E not in scope in type namespace
+     ) where
```
Can we also have a `TypeData` example here? I assume that if I have `type data K = T1`, then I must export/import it as `K(type T1)` rather than `K(data T1)`.

adamgundry replied:
Good idea! Yes my reading of #106 is that the names are introduced in the type namespace, so would use `type` rather than `data`.

adamgundry replied:
I added `TypeData` to Effects and Interactions (https://github.com/adamgundry/ghc-proposals/blob/namespace-specified-imports/proposals/0000-namespace-specified-imports.rst#41typedata).

simonpj commented:
Looks good to me. One nit:
```
import qualified Data.Proxy as D (Proxy) -- Accepted (imports data constructor qualified)
```
Is that right?  It would import the *type* constructor `Proxy` wouldn't it?

Actually for all these Proxy examples it would help to remind us of the declaration of `Proxy` (which uses punning).

I think that
```
  import M type( T(..) )
```
where `T` is an algebraic data constructor, would import just `T` and not the data constructors.  (Possibly with a warning about a misleading `(..)`.

I assume that
```
  import M type ( T( K1, K2 ) )
```
would be an error because it explicitly imports K1, K2, which are in the data space?

Giving these as exmples would be useful.  And checking that the spec covers it.

adamgundry committed:
Flesh out specification, add more examples, discuss interaction with TypeData

jvanbruegge commented:
I find it very weird that
```haskell
import Foo (type T(..))
```
also importa the data constructors. I would expect that it only exports the promoted constructors if using `DataKinds` and warn otherwise. Same for exports. Kind of like TypeData but for usage side, not definition

adamgundry commented:
@jvanbruegge that's already the case, though, and changing it sounds like it could lead to rather painful breakage. I think it makes sense if we understand the namespace specifier as binding tightly, i.e. `import Foo ( (type T) (..) )` rather than `import Foo ( type (T(..)) )`.

jvanbruegge commented:
That's very unfortunate. However a quick look at hackage shows only 5 packages that might be affected: https://hackage-search.serokell.io/?q=import.*%5C%28type+%5B%5E+%5D%2B%5C%28.%2B%5C%29%5C%29

So this might be something for consideration

goldfirere commented on proposals/0000-namespace-specified-imports.rst:
```diff
@@ -0,0 +1,656 @@
+
+* ``import M (data D(..))`` is syntactically valid, but not useful, as it is not
+  currently possible for identifiers in the data namespace to have children.
+  (We might imagine changing this e.g. for pattern synonym record fields, but
```
Or regular data constructor record fields...

goldfirere commented:
Thanks for writing this up! I'm in support.

adamgundry commented:
Discussion on this one seems to have quietened down, so I'd like to submit it to the committee, @nomeata.

angerman commented:
As I've now been assigned Shepard for this, I've given this a few reads and here are my thoughts so far. The overall idea is great, I'm just having trouble with the syntax. Unless I'm mistaking this, permits to write the following to import _only_ the types (type namespace) from `Data.Proxy`
```
import Data.Proxy type
```
However, I find the postfix notation very hard to read; it feels to me like a half finished sentence. And this The intention seems to be to import _all_ type**s** from that module. 
```
import types Data.Proxy
```
to me appears much more obvious what it's going to do. This of course opens the issue of what does
```
import data Data.Proxy
```
mean? And here `values` seems a more appropriate wording, which we appear to want to get away from. From an aesthetics point, I like that type and data are same length, yet type and value seem more readable to me? Maybe I'm the only one.

nomeata commented:
I think the position in the proposal is motivated by `import Foo (explicit, list, of, items)`. But I should let the author answer.

angerman commented:
I don't want to bikeshed this too much. I only want to highlight my usability concerns. My primary focus here is on trying to make sure people _guess_ the right meaning most of the time, even if they are unfamiliar with much of haskell so far.

The example syntax given prominently in the proposal is:
```
import qualified Data.Proxy type as T   -- import only the type namespace
import qualified Data.Proxy data as D   -- import only the data namespace
```
and
```
import qualified Data.Proxy as T (type Proxy)   -- import only the Proxy type
import qualified Data.Proxy as D (data Proxy)   -- import only the Proxy constructor
```

with the change to the grammar being:
```
importdecl -> 'import' [src] ['safe'] ['qualified'] [package] modid [namespace] ['qualified'] ['as' modid] [impspec]

namespace  -> 'data'
           |  'type'
```

which again, would permit writing
```
import X type
```
(which to me is not clear what that's supposed to mean).  After reading the proposal, I know this is supposed to mean "import the type namespace (that is all types defined in module X) from module X", but the connection between the statement, and the meaning isn't very close IMO.

```
import X (type Proxy)
```
on the other hand reads ok to me.

As `ImportQualifiedPost` was brought up, this now ends up reading like
```
import X type qualified as Y
```
what is _type qualified_? And still _type_ is singular, yet I'm potentially getting many types imported.

Regarding the use in explicit import lists
```
import Data.Proxy type (Proxy(..))  -- Accepted: imports type constructor but not data constructor
import Data.Proxy (type Proxy(..))  -- Accepted: imports both type constructor and data constructor
```
seems like very fiddly :-/ I'm just moving the paren a few characters, and all of a sudden I'm additionally importing a data constructor, or not. I'm not sure how easy this is to spot during code reviews.

eyeinsky commented on proposals/0000-namespace-specified-imports.rst:
```diff
@@ -0,0 +1,656 @@
+Where an import has a namespace specifier, any occurrence of an ellipsis
+``(..)`` will be taken to refer only to identifiers in that namespace.  For
+example, ``import M type (T(..))`` will import only type-level names (so the
+ellipsis will refer to nothing if ``T`` is a normal algebraic datatype).
```
The ellipsis will refer to nothing, but the `T` type will still be imported, right? (Maybe say: "so for a normal algebraic data type `T` only the type `T` is imported and ellipsis will refer to nothing")

int-index commented:
@angerman What about this syntax:

```
import Data.Proxy (type ..) as T
import Data.Proxy (data ..) as D
```

The idea is to extend `qcname_ext` with two new forms:

```
  | 'type' '..'
  | 'data' '..'
```

which stand for "all types" and "all data".

michaelpj commented:
If we have `import Data.Proxy (type ...) as T` I would expect to be able to write `import Data.Proxy (...) as T` also! And what about `import Data.Proxy as T hiding (data ...)`?

int-index commented:
@michaelpj I don't know if you wrote `...` instead of `..` on purpose, but we must stick to the latter because

1. `..` is reserved syntax, `...` is a user-defined operator
2. `..` is already used in export/import items, e.g. `import Prelude(Bool(..))`

As to your actual suggestions, they look reasonable to me (i.e. they'd make the proposed syntax more regular), but I can't come up with an actual use case for them, since there are simpler ways of writing both of them.
1. `import Data.Proxy (..) as T` is equivalent to `import Data.Proxy as T`, i.e. just remove the `(..)` altogether
2. `import Data.Proxy as T hiding (data ..)` is equivalent to `import Data.Proxy (type ..) as T`, i.e. use the opposite namespace specifier

My preference would be to omit those unnecessary additions to keep the proposal simple, but ultimately it's up to @adamgundry and I don't feel strongly either way.

angerman commented:
@int-index that doesn't look too bad, and would be mostly in line with what we already have, reusing `..` as meaning wildcard.

```
import Data.Proxy as T (type ..) 
import qualified Data.Proxy as T (type ..)
import Data.Proxy qualified as T (type ..)
```

What about these two now:
```
import Data.Proxy type (Proxy(..))  -- Accepted: imports type constructor but not data constructor
import Data.Proxy (type Proxy(..))  -- Accepted: imports both type constructor and data constructor
```
so we now have
```
import Data.Proxy (     Proxy    )
import Data.Proxy (     Proxy(..))
import Data.Proxy (type ..       )
import Data.Proxy (data ..       )
import Data.Proxy (type Proxy    )  -- Accepted: imports type constructor but not data constructor
import Data.Proxy (type Proxy(..))  -- Accepted: imports both type constructor and data constructor
```
only?

Sorry if I'm getting a bit confused with your proposed syntax, the import one from the proposal.

int-index commented:
@angerman Here's the breakdown.

Today we have:

```
import Data.Proxy   -- import everything

import Data.Proxy (Proxy)          -- 1. import the type constructor only
import Data.Proxy (type Proxy)     -- 2. import the type constructor only
import Data.Proxy (pattern Proxy)  -- 3. import the data constructor only

import Data.Proxy (Proxy(Proxy))       -- 4. import both the type and data constructor
import Data.Proxy (Proxy(..))          -- 5. import both the type and data constructor
import Data.Proxy (type Proxy(Proxy))  -- 6. import both the type and data constructor
import Data.Proxy (type Proxy(..))     -- 7. import both the type and data constructor
```

My proposal is:
1. Replace (3) with `import Data.Proxy (data Proxy)`, i.e. use the `data` keyword instead of `pattern`
2. Allow these two new forms:
    ```
    import Data.Proxy (type ..)    -- 8. import all type constructors
    import Data.Proxy (data ..)    -- 9. import all data constructors and value bindings
    ```

All 9 forms also interact with `qualified` and `as Alias`, but those interactions aren't new and really are orthogonal to the proposed change.

int-index commented:
OK, so to list all the interactions, we end up with these combinations:

```
import qualified X (type ..) as Y
import qualified X (data ..) as Y
import X (type ..) qualified as Y
import X (data ..) qualified as Y
import X (type ..) as Y
import X (data ..) as Y
import qualified X (type ..)
import qualified X (data ..)
import X (type ..) qualified
import X (data ..) qualified
import X (type ..)
import X (data ..)
```

With @michaelpj's first suggestion, we also get

```
import qualified X (..) as Y
import X (..) qualified as Y
import X (..) as Y
import qualified X (..)
import X (..) qualified
import X (..)
```

And with @michaelpj's second suggestion, we also get

```
import qualified X as Y hiding (type ..)
import qualified X as Y hiding (data ..)
import X qualified as Y hiding (type ..)
import X qualified as Y hiding (data ..)
import X as Y hiding (type ..)
import X as Y hiding (data ..)
import qualified X hiding (type ..)
import qualified X hiding (data ..)
import X qualified hiding (type ..)
import X qualified hiding (data ..)
import X hiding (type ..)
import X hiding (data ..)
```

angerman commented:
@adamgundry @Hithroc @cdornan as the proposal authors, what's your take on this?

adamgundry commented:
I like the idea, thanks for the suggestion @int-index! In particular it gets rid of the subtle distinction between `import Data.Proxy (type Proxy(..))` and `import Data.Proxy type (Proxy(..))`, and indeed various corner cases like `import Data.Proxy data (data Proxy)` which are just redundant. Moreover `import Data.Proxy (type ..)` should hopefully be reasonably easy to understand for users who haven't seen it before.

I would be inclined not to permit `import M (..)` because it is entirely redundant. The implementation could always suggest `import M` in the error message. On the other hand, `import M hiding (type ..)` seems relatively natural and might be useful in some circumstances, so I'd suggest allowing it.

This change does blur the notion of an explicit import list or explicit export list. Presumably `-Wmissing-import-lists` would not fire for `import M (type .., data ..)` even though it is essentially equivalent to `import M` (and similarly for `-Wmissing-export-lists`). I think this is okay though.

I'm happy to rewrite the proposal to use this syntax, if we have consensus that this is the right approach. How do others feel about it?

Hithroc commented:
I'm okay with the proposed syntax.

adamgundry committed:
Revise namespace-specified imports proposal

This changes the proposal to use

  import M (type ..)

instead of

  import M type

thanks to a suggestion from Vladislav Zavialov.

adamgundry commented:
I've now revised the proposal to take account of the new syntax, and I'm very happy with the new design. Thanks @int-index for the suggestion! A few things occurred to me while doing so:

* It is now natural to permit `type ..` or `data ..` to appear as a child import item, for example:
   ```hs
   import M (C (type ..))
   ```
   This might be useful if `C` is a class with many children in both namespaces (e.g. both associated types and methods). Thus I'm inclined to allow it.

* We can now export all names in a single namespace:  
   ```hs
   module M (type ..) where ...
    ```

* We can now somewhat redundantly import both namespaces:
   ```hs
   import M (type.., data ..)
   ```

phadej commented:
> Presumably `-Wmissing-import-lists` would not fire for `import M (type .., data ..)`

I'd suggest keep it firing for these. The very important point of `-Wmissing-import-lists` is to find imports which may cause problems if dependencies add new bindings. The way to silence `-Wmissing-import-lists` should not be to change `import M` into `import M(data ..)`, but to spell the imports (or import qualified).
